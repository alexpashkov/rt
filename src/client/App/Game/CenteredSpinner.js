import React from 'react';
import { PacmanLoader } from 'react-spinners';

const CenteredSpinner = () => (
  <div
    css={`
      display: flex;
      align-items: center;
      justify-content: center;
      width: 100%;
    `}
  >
    <PacmanLoader color="red" />
  </div>
);

export default CenteredSpinner;
