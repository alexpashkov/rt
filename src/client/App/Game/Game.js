/* eslint react/prop-types: 0 */ // prop-actionTypes are set in the container component
import React from 'react';
import GameBoard from './GameBoard/GameBoard';

const Game = ({ userBoard = null }) => <GameBoard board={userBoard} />;

export default Game;
