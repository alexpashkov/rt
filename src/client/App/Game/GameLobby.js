import React from "react";

const GameLobby = () => <div>Game Lobby</div>;

export default GameLobby;