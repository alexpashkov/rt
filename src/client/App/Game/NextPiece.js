import React from "react";
import Board from "./Board/Board";
import {numToBinary16String} from "../../utils";
import * as R from "ramda";

const appendPrepend = R.curry((v, coll) => R.compose(R.append(v), R.prepend(v))(coll));
const emptyRow = R.range(0, 6).map(R.always(0));

const getChunks = R.curry((n, coll) => {
  let i = 0;
  const chunks = [];
  while (coll && i < coll.length) {
    chunks.push(coll.slice(i, i + n));
    i += n;
  }
  return chunks;
});

const pieceToBoard = (piece) => {
  return R.ifElse(
    Boolean,
    R.compose(
      R.map(row =>  row.map(Number)),
      getChunks(4),
      R.split(""),
      numToBinary16String
    ),
    R.always(
      [
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
      ]
    )
  )(piece);
}

// eslint-disable-next-line
export default ({piece}) => {
  const board = pieceToBoard(piece);
  return <div css={`
    padding: 10px;
  `}>
    <h2>Next piece</h2>
    <Board board={board}/></div>;
};
