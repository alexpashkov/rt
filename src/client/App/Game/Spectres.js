import React from 'react';
import styled from 'react-emotion';
import Board from './Board/Board';

const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: flex-start;
  color: ${({ theme }) => theme.colors.secondaryText};
`;

const SpectreWrapper = styled.div`
  width: 50%;
  padding: 5px;
`;

export default ({ spectres = [] }) => (
  <Wrapper>
    {spectres.length
      ? spectres.sort((a, b) => a.id - b.id).map(({ spectre, id }, i) => (
          <SpectreWrapper key={id}>
            <h2 css={`
              font-size: 1rem;
              overflow: hidden;
              text-overflow: ellipsis;
            `}>{id}</h2>
            <Board board={spectre} />
          </SpectreWrapper>
        ))
      : 'You are alone'}
  </Wrapper>
);
