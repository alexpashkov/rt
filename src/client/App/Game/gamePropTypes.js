import PropTypes from 'prop-types';

export default {
  userId: PropTypes.string,
  currentGameInfo: PropTypes.shape({
    id: PropTypes.string.isRequired,
    isRunning: PropTypes.bool.isRequired,
    leaderId: PropTypes.string.isRequired,
    players: PropTypes.arrayOf(
      PropTypes.shape({
        login: PropTypes.string.isRequired
      })
    ).isRequired,
    chatHistory: PropTypes.array.isRequired
  }),
  gameId: PropTypes.string.isRequired
}
