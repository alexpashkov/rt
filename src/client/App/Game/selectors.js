import { createSelector } from 'reselect';
import { numToBinary16String } from '../../utils';
import { pickBy } from 'ramda';

export const mergeBoardAndPiece = (board, piece) => {
  if (!piece) return board;

  const pieceCells = getPieceCells(piece.code);
  return board.map(
    (row, y) =>
      pieceCells.find(cell => cell.y + piece.y === y)
        ? row.map(
            (cell, x) =>
              pieceCells.find(
                cell => cell.x + piece.x === x && cell.y + piece.y === y
              )
                ? piece.code
                : board[y][x]
          )
        : row
  );
};

export const getPieceCells = pieceCode =>
  numToBinary16String(pieceCode)
    .split('')
    .reduce(
      (acc, curr, i) =>
        +curr
          ? [
              ...acc,
              {
                x: i % 4,
                y: Math.floor(i / 4)
              }
            ]
          : acc,
      []
    );

export const userBoardSelector = createSelector(
  state => (state.boards && state.user && state.boards[state.user.id]) || [],
  state => state.currentPiece,
  mergeBoardAndPiece
);

export const spectresSelector = createSelector(
  state => state.boards || {},
  state => state.user && state.user.id,
  (boards, userId) =>
    boards && userId
      ? Object.keys(pickBy((_, id) => id != userId, boards))
          .reduce(
            (spectres, id) => [
              ...spectres,
              Object.assign({
                id,
                spectre: spectreFromBoard(boards[id])
              })
            ],
            []
          )
          .sort()
      : []
);

export const spectreFromBoard = board => {
  const spectre = Array.from(Array(20)).map(() => Array.from(Array(10)));
  for (let x = 0; x < 10; x++) {
    let filledFound = 0;
    for (let y = 0; y < 20; y++) {
      spectre[y][x] = (filledFound = filledFound || board[y][x]) ? 1 : 0;
    }
  }
  return spectre;
};
