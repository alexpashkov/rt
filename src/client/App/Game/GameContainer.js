import {
  compose,
  branch,
  lifecycle,
  renderComponent,
  withProps,
  setPropTypes
} from 'recompose';
import { connect } from 'react-redux';

import gamePropTypes from './gamePropTypes';
import { setInfo as setCurrentGameInfo } from '../../actions/currentGameInfoActions';
import history from '../../history';
import socket from '../../socket';
import CenteredSpinner from './CenteredSpinner';
import GameLobby from './GameLobby';
import Game from './Game';
import { setBoard, clearBoards } from '../../actions/boardsActions';
import { setCurrentPiece, setNextPiece } from '../../actions/pieceActions';
import { userBoardSelector, spectresSelector } from './selectors';
import withRunningGameLogic from './withRunningGameLogic';

import {
  client as clientSocketEvents,
  server as serverSocketEvents
} from '../../../shared/socket-events';

const emitGameJoin = (gameId, cb) =>
  socket.emit(clientSocketEvents.ROOM_JOIN, { id: gameId }, cb);
const emitGameLeave = gameId =>
  socket.emit(clientSocketEvents.ROOM_LEAVE, { id: gameId });

export default compose(
  connect(
    state => ({
      userId: state.user && state.user.id,
      userBoard: userBoardSelector(state),
      spectres: spectresSelector(state),
      currentGameInfo: state.currentGameInfo,
      nextPiece: state.piece.next
    }),
    {
      setCurrentGameInfo,
      setBoard,
      clearBoards,
      setCurrentPiece,
      setNextPiece
    }
  ),
  withProps(({ match: { params: { gameId } } }) => ({
    gameId
  })),
  lifecycle({
    componentDidMount() {
      const {
        gameId,
        setCurrentGameInfo,
        setCurrentPiece,
        setNextPiece,
        setBoard
      } = this.props;
      emitGameJoin(gameId, ({ status, roomInfo }) => {
        if (status === 'error') {
          /* failed to join the game, redirect to lobby */
          // alert(description || 'Failed to join the game');
          return history.push('/');
        }
        setCurrentGameInfo(roomInfo);
      });
      /* subscribe to game info updates, e.g when new player joins the game
      is reflected for players who's in the game lobby */
      socket.on(serverSocketEvents.ROOM_INFO_UPDATE, setCurrentGameInfo);
      socket.on(serverSocketEvents.GAME_PIECE_CURRENT, ({ piece }) =>
        setCurrentPiece(piece)
      );
      socket.on(serverSocketEvents.GAME_PIECE_NEXT, data => {
        setNextPiece(data.piece);
      });
      socket.on(serverSocketEvents.GAME_BOARD_CURRENT, data =>
        setBoard(data.id, data.board)
      );
      socket.on(serverSocketEvents.GAME_FINISHED, ({ winnerId = '' } = {}) => {
        alert(
          winnerId ? 'Game finished!\nWinner is ' + winnerId : "You've lost!"
        );
      });
    },
    componentWillUnmount() {
      const { gameId, setCurrentGameInfo } = this.props;
      /* emit leave event for the server */
      emitGameLeave(gameId);
      /* set current game info to null, so we can use it as indicator for a spinner */
      setCurrentGameInfo(null);
      socket.off(serverSocketEvents.ROOM_INFO_UPDATE);
      socket.off(serverSocketEvents.GAME_PIECE_CURRENT);
      socket.off(serverSocketEvents.GAME_BOARD_CURRENT);
      socket.off(serverSocketEvents.GAME_FINISHED);
    }
  }),
  setPropTypes(gamePropTypes),
  branch(
    ({ currentGameInfo }) => !currentGameInfo,
    renderComponent(CenteredSpinner)
  ),
  branch(
    ({ currentGameInfo }) => !currentGameInfo.isRunning,
    renderComponent(GameLobby)
  ),
  withRunningGameLogic
)(Game);
