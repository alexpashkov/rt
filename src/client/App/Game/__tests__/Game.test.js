import React from 'react';
import renderer from 'react-test-renderer';
import Game from '../Game';
import { ThemeProvider } from '../../ThemeProvider';

describe('Game', () => {
  it('Renders', () => {
    const component = renderer.create(
      <ThemeProvider>
        <Game />
      </ThemeProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
