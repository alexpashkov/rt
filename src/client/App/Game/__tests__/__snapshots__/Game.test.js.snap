// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Renders 1`] = `
<div
  className="css-zdcrpv e1j8727c0"
>
  <div
    className="css-wmnluo e1j8727c2"
  />
  <div
    className="css-1ap4i72 e1j8727c3"
  >
    <div
      className="css-bkjqz6 e9pbz0x0"
    >
      You are alone
    </div>
  </div>
</div>
`;
